### PR TITLE
gh-92536: PEP 623: Remove wstr from unicode

### PR DESCRIPTION
**Pull request by methane**
gh-92536

### CODE REVIEW

File: Doc/c-api/unicode.rst [hunk 1]
```diff
@@ -85,6 +85,16 @@ access to internal read-only data of Unicode objects:
...
 
 
+.. c:function:: int PyUnicode_READY(PyObject *o)
+
+   Returns ``0``. This API is kept only for backward compatibility.
+
+   .. versionadded:: 3.3
+
+   .. deprecated:: 3.10
+      This API do nothing since Python 3.12. Please remove code using this function.
```
Comment (methane) [inline Doc/c-api/unicode.rst:95]: Huge number of code using this API.
https://github.com/hpyproject/top4000-pypi-packages/search?q=PyUnicode_READY

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -91,7 +91,9 @@ Improved Modules
 Optimizations
 =============
 
-
+* Removed ``wstr`` and ``wstr_length`` members from Unicode objects.
+  It reduces object size 8 or 16 bytes on 64bit platform.
```
Comment (arhadthedev) [inline Doc/whatsnew/3.12.rst]: Probably it should be:

```suggestion
  It reduces object size by 8 or 16 bytes on 64bit platform.
```

**Comment by methane** [comment]
I added @serhiy-storchaka to reviewer because he is author of USE_UNICODE_WCHAR_CACHE.

**Review by ericsnowcurrently — COMMENTED** [review]
Nothing wrong stood out after a quick scan on my phone.  The change seems mostly mechanical and limited to removals.  Is that right?

Also, should the PEP be mentioned more prominently in the Misc/NEWS entry and whats_new?

**Comment by serhiy-storchaka** [comment]
What if merge #12409 first and set HAVE_UNICODE_WCHAR_CACHE to 0? It will have an effect of removing wstr, but keep possibility to return it in custom builds.

**Comment by arhadthedev** [comment]
[Argument Clinic How-To](https://docs.python.org/3/howto/clinic.html) also needs an update:

```diff
diff --git a/Doc/howto/clinic.rst b/Doc/howto/clinic.rst
index 04b1a2cac0..f25f38c0a5 100644
--- a/Doc/howto/clinic.rst
+++ b/Doc/howto/clinic.rst
@@ -855,8 +855,6 @@ on the right is the text you'd replace it with.
 ``'y'``     ``str(accept={bytes})``
 ``'y#'``    ``str(accept={robuffer}, zeroes=True)``
 ``'y*'``    ``Py_buffer``
-``'Z'``     ``Py_UNICODE(accept={str, NoneType})``
-``'Z#'``    ``Py_UNICODE(accept={str, NoneType}, zeroes=True)``
 ``'z'``     ``str(accept={str, NoneType})``
 ``'z#'``    ``str(accept={str, NoneType}, zeroes=True)``
 ``'z*'``    ``Py_buffer(accept={buffer, str, NoneType})``
```

**Comment by methane** [comment]
> What if merge #12409 first and set HAVE_UNICODE_WCHAR_CACHE to 0? It will have an effect of removing wstr, but keep possibility to return it in custom builds.

I think it is very difficult to maintain both of HAVE_UNICODE_WCHAR_CACHE is 0 and 1.



> [Argument Clinic How-To](https://docs.python.org/3/howto/clinic.html) also needs an update:

AC supports u and Z even after removing wstr. I will replace Py_UNICODE with wchar_t in the clinic howto.

File: Doc/c-api/arg.rst [hunk 3]
```diff
@@ -247,6 +205,11 @@ which disallows mutable objects such as :class:`bytearray`.
    them. Instead, the implementation assumes that the byte string object uses the
    encoding passed in as parameter.
 
+.. versionchanged:: 3.12
+   ``u``, ``u#``, ``Z``, and ``Z#`` are removed because they used legacy ``Py_UNICODE*``
```
Comment (slateny) [inline Doc/c-api/arg.rst:209]: ... because they used `a/the` legacy ...

File: Doc/c-api/unicode.rst (hunk 1 above)
Comment (slateny) [reply Doc/c-api/unicode.rst:95]: This API `does` nothing..., and I think the 'Please remove ...' can be omitted

**Review by slateny — COMMENTED** [review]
some optional grammar changes if you revisit some of these files later down the line

File: Doc/c-api/arg.rst (hunk 3 above)
Comment (methane) [reply Doc/c-api/arg.rst:209]: Thank you. I made follow-up PR: https://github.com/python/cpython/pull/92756

**Comment by methane** [comment]
> Congratulations for finally managing to remove this legacy API. A big **thank you**! I hated this API since Python 3.3, having to call "PyUnicode_Ready()" was super annoying. Thanks for you tenacity to remove it!

I'm sorry but I am considering to un-deprecate `PyUnicode_READY()`.
Please visit this thread.
https://discuss.python.org/t/undeprecate-pyunicode-ready-for-future-unicode-improvement/15717

**Review by ameenkhan003 — COMMENTED** [review]
hytt

**Pull request merged**